src/Workspaces/Core/Portable/Shared/Extensions/IAsyncEnumerableExtensions.cs

```diff
@@ -20,5 +34,52 @@ await foreach (var value in values.WithCancellation(cancellationToken).Configure
 
             return result.ToImmutable();
         }
+
+        /// <summary>
+        /// Takes an array of <see cref="IAsyncEnumerable{T}"/>s and produces a single resultant <see
+        /// cref="IAsyncEnumerable{T}"/> with all their values merged together.  Absolutely no ordering guarantee is
+        /// provided.  It will be expected that the individual values from distinct enumerables will be interleaved
+        /// together.
+        /// </summary>
+        /// <remarks>This helper is useful when doign parallel processing of work where each job returns an <see
+        /// cref="IAsyncEnumerable{T}"/>, but one final stream is desired as the result.</remarks>
+        public static IAsyncEnumerable<T> MergeAsync<T>(this ImmutableArray<IAsyncEnumerable<T>> streams, CancellationToken cancellationToken)
+        {
+            // Code provided by Stephen Toub, but heavily modified after that.
+
+            // 1024 chosen as a way to ensure we don't necessarily create a huge unbounded channel, while also making it
+            // so that we're unlikely to throttle on any stream unless there is truly a huge amount of results in it.
+            var channel = Channel.CreateBounded<T>(1024);
+
+            var tasks = new Task[streams.Length];
+            for (var i = 0; i < streams.Length; i++)
+                tasks[i] = Process(streams[i], channel.Writer, cancellationToken);
+
+            // Complete the channel writer with the result of all the tasks.  If nothing failed, t.Exception will be
+            // null and this will complete successfully.  If anything failed, the exception will propagate out.
+            //
+            // Note: passing CancellationToken.None here is intentional/correct.  We must complete all the channels to
+            // allow reading to complete as well.
```

Comment:
How does cancellation work? In the cancelled case, is the exception here an OperationCancelledException, or something else? And since Exception here actually returns an _aggregate_ exception, is that going to cause us to wrap exceptions when we don't expect it? Or does .Complete() here do unwrapping of the aggregate exception?

Reply:
this is a good question.  i'll wait on @stephentoub to explain this because i'm not actually sure if this handled cancellation properly.

From muy reading, you'll get an AggregateException wrapping all the cancellation exceptions.  i *think* the TPL sees through this when completing the writer (and then exposing that on the reader side), so you hopefully just get cancellation on the reading side.

But i'm not sure.  @stephentoub for insights.

Reply:
i just debugged this with a little test.  The task you get here is:

![image](https://user-images.githubusercontent.com/4564579/196301733-20cf7ea6-2c21-4bcf-a0e9-345adfc6588e.png)


Note that it is in the canceled state, but `.Exception` is null.  From teh docs, this is what is said:

> If none of the supplied tasks faulted but at least one of them was canceled, the returned task will end in the [Canceled](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskstatus?view=net-6.0#system-threading-tasks-taskstatus-canceled) state.

> If you're waiting on a Task that transitions to the Canceled state, a [System.Threading.Tasks.TaskCanceledException](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskcanceledexception) exception (wrapped in an [AggregateException](https://learn.microsoft.com/en-us/dotnet/api/system.aggregateexception) exception) is thrown. This exception indicates successful cancellation instead of a faulty situation. Therefore, the task's [Exception](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.exception) property returns null.

The last line is particularly relevant.

So it seems like task.Exception is null in this case.  This is also something you can trivially demonstrate by doing Task.FromCanceled(canceledToken) and checking .Exception on that.

So i think we're all good here.  Will comment in next pr.